orttraining/orttraining/training_api/module.cc

```diff
@@ -48,62 +49,103 @@ Status Parameter::ResetGrad() {
   train_input_names_.insert(train_input_names_.end(), grad_input_names.begin(), grad_input_names.end());
   train_input_names_.insert(train_input_names_.end(), reset_grad_name.begin(), reset_grad_name.end());
 
-  // Eval model validation
-  if (nullptr != eval_sess_) {
+  // Loop each parameter, allocate it's memory based on user specified device.
+  for (auto& param_name : param_input_names) {
+    ORT_ENFORCE(named_parameters.find(param_name) != named_parameters.end());
+    OrtValue& source_ortvalue = named_parameters[param_name]->Data();
+    ORT_ENFORCE(source_ortvalue.IsTensor());
+    const Tensor& source_tensor = source_ortvalue.Get<Tensor>();
+
+    std::vector<SessionState::NodeInfo> node_info_vec;
+    ORT_THROW_IF_ERROR(train_sess_state.GetInputNodeInfo(param_name, node_info_vec));
+    const auto& node_info = node_info_vec.front();
+    const auto target_device = *node_info.device;
+    for (auto it = node_info_vec.begin(); it != node_info_vec.end(); ++it) {
+      ORT_ENFORCE(target_device == *(it->device));
+    }
+
+    // Create parameter value copy with corresponding device user sets the session on.
+    // We did not re-use the data even CPU tensor is needed.
+    // TODO(pengwa): consider whether we should alloc contiguous buffer for parameters or gradients.
+    OrtValue target_ortvalue;
+    auto allocator = train_sess_state.GetAllocator(target_device);
+    ORT_ENFORCE(allocator != nullptr);
+
+    Tensor::InitOrtValue(source_tensor.DataType(),
+                         source_tensor.Shape(),
+                         allocator, target_ortvalue);
+    Tensor* target_tensor_ptr = target_ortvalue.GetMutable<Tensor>();
+    ORT_THROW_IF_ERROR(train_sess_state.GetDataTransferMgr().CopyTensor(source_tensor, *target_tensor_ptr));
+
+    auto param_share_ptr = std::make_shared<Parameter>(param_name, target_ortvalue, named_parameters[param_name]->RequiresGrad());
+    named_parameters_.insert({param_name, param_share_ptr});
+    weights_.push_back(param_share_ptr->Data());
+
+    // Create gradient buffer when paramter requires gradient.
+    if (param_share_ptr->RequiresGrad()) {
+      // Create gradient accumulation buffer.
+      auto it = param_name_to_grad_input_index_map.find(param_name);
+      ORT_ENFORCE(it != param_name_to_grad_input_index_map.end(), "Gradient buffer input not providered for param: ", param_name);
+
+      auto& param_grad_buffer_name = grad_input_names[param_name_to_grad_input_index_map[param_name]];
+      // TODO: don't pre-allocate the gradient buffer.
+      // Gradient usually stays on the same device of its parameter.
+      OrtValue param_grad_buffer_ortvalue;
+      ORT_THROW_IF_ERROR(utils::OrtValueLike(train_sess_state, target_ortvalue, param_grad_buffer_ortvalue));
+      ORT_THROW_IF_ERROR(param_share_ptr->AllocateGrad(param_grad_buffer_name, param_grad_buffer_ortvalue));
+      gradients_[param_name_to_grad_input_index_map[param_name]] = param_share_ptr->Gradient();
+    }
+  }
+}
```

Comment:
This constructor seems to be doing too much different work. We should use functions in an unnamed namespace that the constructor can call into.

Reply:
True, but let me do that in a new PR.  don't want to block you guys with this. :)